### ISSUE TITLE
oc cluster up reports "getsockopt: connection refused"

### ISSUE THREAD

**Issue by slaskawi** [openshift/origin#17103]
When running `oc cluster up --service-catalog` (it also happens without `--service-catalog`), I got the following error:

```
$ oc cluster up --service-catalog --server-loglevel=5
Starting OpenShift using openshift/origin:v3.6.0 ...
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... 
   Deleted existing OpenShift container
-- Checking for openshift/origin:v3.6.0 image ... OK
-- Checking Docker daemon configuration ... OK
-- Checking for available ports ... OK
-- Checking type of volume mount ... 
   Using nsenter mounter for OpenShift volumes
-- Creating host directories ... OK
-- Finding server IP ... 
   Using 127.0.0.1 as the server IP
-- Checking service catalog version requirements ... OK
-- Starting OpenShift container ... 
   Creating initial OpenShift configuration
   Starting OpenShift using container 'origin'
   Waiting for API server to start listening
FAIL
   Error: cannot access master readiness URL https://127.0.0.1:8443/healthz/ready
   Details:
     No log available from "origin" container

   Caused By:
     Error: Get https://127.0.0.1:8443/healthz/ready: dial tcp 127.0.0.1:8443: getsockopt: connection refused
```

Logs from the server might be found here: https://gist.github.com/slaskawi/ba7d3f11228c4e0aec964fb43a9ad092

##### Version

```
$ oc version
oc v3.6.0+c4dd4cf
kubernetes v1.6.1+5115d708d7
features: Basic-Auth GSSAPI Kerberos SPNEGO

$ docker version
Client:
 Version:         1.12.6
 API version:     1.24
 Package version: docker-1.12.6-61.git85d7426.el7.centos.x86_64
 Go version:      go1.8.3
 Git commit:      85d7426/1.12.6
 Built:           Tue Oct 24 15:40:21 2017
 OS/Arch:         linux/amd64

Server:
 Version:         1.12.6
 API version:     1.24
 Package version: docker-1.12.6-61.git85d7426.el7.centos.x86_64
 Go version:      go1.8.3
 Git commit:      85d7426/1.12.6
 Built:           Tue Oct 24 15:40:21 2017
 OS/Arch:         linux/amd64
```

##### Additional Information
It is worth to mention that this happen on a pretty slow CI. It's pretty hard to tell since all our servers in CI installation are the same but this one seems be "slower" than the rest. It's very subjective, but still, it may give you some hints.

**Comment by xiaods** [comment]
not found in local dev env.

**Comment by slaskawi** [comment]
@xiaods Works on my local dev box also. The thing is that it seems not to work correctly in a constrained environment.

**Comment by pweil-** [comment]
@mfojtik in the spirit of making sure that the `cluster-up` work is distributed to the right owners I'm throwing this your way.  `cluster-up` encompasses all components of the cluster and is getting too much for one person to handle.  

This looks like a timeout waiting on master to start perhaps.

/cc @csrwng

**Comment by csrwng** [comment]
@slaskawi what do you mean by constrained environment? Is this inside a vm? can you include hardware/OS details ?

**Comment by slaskawi** [comment]
@csrwng Sure! It's one of our internal RH VMs (hosted on OpenStack) with 4 GB of RAM with 2 vCPUs. If you asked me offline I can grant you access to it.

**Comment by z0x010** [comment]
When running `oc cluster up --public-hostname 127.0.0.1` on my MacBook Pro, I have got this error:
```
jack@mbp $ oc cluster up --public-hostname 127.0.0.1  
Starting OpenShift using openshift/origin:v3.6.1 ...
-- Checking OpenShift client ... OK
-- Checking Docker client ... OK
-- Checking Docker version ... OK
-- Checking for existing OpenShift container ... OK
-- Checking for openshift/origin:v3.6.1 image ... OK
-- Checking Docker daemon configuration ... OK
-- Checking for available ports ... OK
-- Checking type of volume mount ... 
   Using Docker shared volumes for OpenShift volumes
-- Creating host directories ... OK
-- Finding server IP ... 
   Using public hostname IP 127.0.0.1 as the host IP
   Using 127.0.0.1 as the server IP
-- Starting OpenShift container ... 
   Creating initial OpenShift configuration
   Starting OpenShift using container 'origin'
   Waiting for API server to start listening
FAIL
   Error: cannot access master readiness URL https://127.0.0.1:8443/healthz/ready
   Details:
     No log available from "origin" container

   Caused By:
     Error: Get https://127.0.0.1:8443/healthz/ready: EOF
```

### version : 

```
jack@mbp $ oc version
oc v3.6.1+008f2d5
kubernetes v1.6.1+5115d708d7
features: Basic-Auth

----------------------------------
jack@mbp $ docker version
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 08:47:51 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba3
 Built:        Wed Feb  8 08:47:51 2017
 OS/Arch:      linux/amd64
 Experimental: true

```

**Comment by slaskawi** [comment]
@csrwng @z0x010 I have seen two variants of this issue so far:

1. Healthcheck was malfunctioning because of the firewall settings. Check what happens when you call `iptables -F` (or equivalent).
2. Not enough memory. Recently I resized our CI environment from 4 GB or RAM to 8. And this actually started to work correctly.

The interesting thing is that openshift container logs suggested that it constantly was making some progress. It was just too slow (it may happen if you swap into a very slow disk). 

So my suggestion to maintainers would be to either expose timeout option or periodically check the logs if the openshift container makes any progress...

**Comment by csrwng** [comment]
@slaskawi thanks, I will look into monitoring progress somehow.

@z0x010 how much memory/cpu are you giving Docker? Also, is socat functional on your machine? (as in does `socat --help` work?)

**Comment by slaskawi** [comment]
@csrwng I also noticed that is still happening from time to time on our environment (even though we increased memory settings). I'm still trying to figure out why this is happening on certain slaves only (the funny thing is that all slaves are configured using Ansible, so everything should be exactly the same).

**Comment by csrwng** [comment]
@slaskawi so I just tried on my Mac, giving Docker 2GB memory, 1 cpu and cluster up still started ok. I'll try on a VirtualBox Linux VM next. I want to understand how we can check whether origin is still progressing so we can exit or continue waiting.

**Comment by slaskawi** [comment]
@csrwng Thanks for the update.

I will be trying to find some pattern in this. Currently our CI is quite busy (tomorrow we plan to release a new version) but let me get back to you at the end of the week or early next week.

**Comment by slaskawi** [comment]
Another portion of logs: https://justpaste.it/1d9jy

```
Nov 7 15:47:32 rhos-infinispan-slave-fast-1 journal: F1107 15:47:32.572095 15552 start_master.go:506] timed out waiting for the condition
```

**Comment by csrwng** [comment]
@slaskawi thanks for the logs. This isn't a cluster up timeout. It's actually the master timing out waiting for controllers to start: https://github.com/openshift/origin/blob/008f2d5528bf998326b5eb3f1fe3144c59392b9d/pkg/cmd/server/start/start_master.go#L506

Can you try invoking cluster up with --server-loglevel=5 to determine why this is happening?

**Comment by csrwng** [comment]
I also started a Fedora 26 vm and assigned it 4GB memory. I haven't seen an issue yet starting cluster up.

**Comment by slaskawi** [comment]
@csrwng Disk I/O - that could be a really good hint! Let me try to investigate this further and get back to you with some results...

**Comment by slaskawi** [comment]
@csrwng Indeed that was the disk! It turned out that we used NFS on our CI environment which is quite slow (I benchmarked it using `sysbench fileio`). Thanks again for a really good pointer.

However I still think it might be a good idea either to expose a timeout option (since the server was still booting up when it received timout; if it had had 1 or 2 more minutes it would have booted up) or check if the server makes any progress before killing it.

But it's actually your call - feel free to either close this ticket or implement some enhancements based on it.

**Comment by csrwng** [comment]
@slaskawi the timeout that would need to be modified is the controller start timeout in the master code (fixed at 30 sec):  
https://github.com/openshift/origin/blob/release-3.6/pkg/cmd/server/start/start_master.go#L723-L739
Given that this has been implemented differently in the latest master (3.7), I don't think we will backport a fix for 3.6.
